### PR TITLE
Npm for vendoring files

### PR DESCRIPTION
**Pull request by jacksonargo**

### CODE REVIEW

**Comment by 0az** [comment]
I'm actually going to strong reject this (-1), in favor of using CDNJS or JSDelivr for now. Once we have a *proper* need for NPM/Yarn, we can transition. However, until then, it's just unnecessary complexity.

The CDN's faster, anyway.

**Comment by jacksonargo** [comment]
No, I'm not distributing anything with npm. Just using it to vendor js packages. I'm confused what the issue is?

**Comment by 0az** [comment]
Basically, source the scripts from:
- https://www.jsdelivr.com/package/npm/jquery
- https://www.jsdelivr.com/package/npm/datatables

We can pin the version, and use SRI for verification.

**Comment by jacksonargo** [comment]
Oh I see! Ok that makes more sense. I'd like to have the vendor file, so that we know all the external packages we're using. That way, if we want to local clients w/ electron or something, we'll already know all our dependencies, instead of grep `jsdeliver` from code. (This also makes local dev super easy, since we're pulling everything locally instead of from jsdelivr's cdn and my ide isn't complaining).

As far as _actually_ delivering the js files in production, we can use a template vars for the path to these libraries.

Would that solve your concerns?

**Comment by 0az** [comment]
> I'd like to have the vendor file, so that we know all the external packages we're using.
At this point, we use exactly two external deps on exactly one page, so I think it's premature to worry about this.

> This also makes local dev super easy, since we're pulling everything locally instead of from jsdelivr's cdn and my ide isn't complaining
For this, manual installation of
- https://www.npmjs.com/package/@types/jquery
- https://www.npmjs.com/package/@types/datatables.net
will do for now.

**Comment by jacksonargo** [comment]
I don't understand the issue? I would really like to vendor the js dependencies here.

**Comment by 0az** [comment]
I guess we can just toss them in the `package.json`. I want to use Yarn though [0], so for now I'll ask that the lockfile stay out of the PR.

> As far as actually delivering the js files in production, we can use a template vars for the path to these libraries.

This is fine for now. How does this sound?

[0]: https://yarnpkg.com/features/pnp is the killer feature.

File: Dockerfile [hunk 1]
```diff
@@ -18,7 +23,8 @@ CMD ["make", "test"]
 FROM gcr.io/distroless/base-debian10 as vvgo
 COPY ./infra/vvgo/etc/mime.types /etc/
 COPY ./public /public
-COPY --from=builder vvgo vvgo
+COPY --from=builder vvgo /vvgo
+COPY --from=node node_modules /node_modules
```
Comment (0az) [inline Dockerfile:27]:
```suggestion
```

File: Dockerfile [hunk 2]
```diff
@@ -1,3 +1,8 @@
+FROM node:13.12.0 as node
+COPY package.json .
+COPY package-lock.json .
+RUN npm install
+
```
Comment (0az) [inline Dockerfile]:
```suggestion
```

File: Dockerfile [hunk 3]
```diff
@@ -1,3 +1,8 @@
+FROM node:13.12.0 as node
+COPY package.json .
+COPY yarn.lock .
+RUN yarn install
+
```
Comment (0az) [inline Dockerfile:5]:
```suggestion
```

File: Dockerfile (hunk 1 above)
Comment (0az) [reply Dockerfile:27]: We shouldn't be including `node_modules` in our final Docker image.

File: public/npm [hunk 4]
```diff
@@ -0,0 +1 @@
+../node_modules/
```
Comment (0az) [inline public/npm:1]: Requesting delete, we should use JSDelivr instead of `node_modules`.

File: public/parts.gohtml [hunk 5]
```diff
@@ -11,11 +11,11 @@
...
 
     <link rel="icon" href="images/favicons/favicon-32x32.png" sizes="32x32" type="image/png">
+    <link rel="stylesheet" type="text/css" href="css/jquery.dataTables.css">
     <link rel="stylesheet" type="text/css" href="css/theme.css">
     <link rel="stylesheet" type="text/css" href="css/parts.css">
-    <link rel="stylesheet" type="text/css" href="css/jquery.dataTables.css">
-    <script type="text/javascript" src="js/jquery.js"></script>
-    <script type="text/javascript" src="js/jquery.dataTables.js"></script>
+    <script type="text/javascript" src="npm/jquery/dist/jquery.min.js"></script>
+    <script type="text/javascript" src="npm/datatables.net/js/jquery.dataTables.min.js"></script>
```
Comment (0az) [inline public/parts.gohtml:18]: Replace with CDN+SRI.

File: public/npm (hunk 4 above)
Comment (jacksonargo) [reply public/npm:1]: Lmao we can serve from jsdelivr and still have this link for local development.

File: Dockerfile (hunk 1 above)
Comment (jacksonargo) [reply Dockerfile:27]: Why cant we include them in the final image?

File: public/npm (hunk 4 above)
Comment (0az) [reply public/npm:1]: IDK, feels too much like exposing internal details. I suppose I'd evaluate it at -.5, but I guess it doesn't really matter.

**Comment by jacksonargo** [comment]
I'm curious about the cdn thing? Why use these cdn's over cloud-flare page caching?

**Comment by jacksonargo** [comment]
Also, isn't kinda like a security thing too right? We want to be serving _our_ application code instead relying on some 3rd party cdn. The yarn.lock file gives us package validation.

**Comment by 0az** [comment]
JsDelivr is marginally faster than Cloudflare, by virtue of serving from multiple CDNs. It also allows for better caching by user agents - get jQuery from the CDN once (very common).

If we wanted, we could use CDNJS instead, which is literally Cloudflare.

However, the main benefit is that we don't have to deal with it, and instead can just `script src=...`.

**Comment by 0az** [comment]
> Also, isn't kinda like a security thing too right? We want to be serving our application code instead relying on some 3rd party cdn. The yarn.lock file gives us package validation.

SRI does that. Example:

**Comment by jacksonargo** [comment]
Yeah.... but then that means I have to go find all the SRI for every file that I want to import. And I DONT want to do that. That sounds like a mega pain.

**Comment by 0az** [comment]
JSDelivr has a "copy HTML with SRI" option, which is what I used for that snip.

**Comment by jacksonargo** [comment]
What happens when we change versions and need to update all the SRI's? Is there a tool to fetch them? Is there some template service that we can use to manage them when they change?

**Comment by 0az** [comment]
We just copy the new version's script tag from JSDelivr. Realistically though, there's a 99.9% chance we won't be bumping the versions.

If we're bumping versions, at that point we can start considering a proper bundling system, such as Rollup (Webpack is a... pain to config, Rollup is what I've been using as a good mid-point).

**Comment by jacksonargo** [comment]
I would WAY prefer to just let yarn handle all that and serve the files from the container. We have nginx and cloudflare caching. Why is that not enough?

**Comment by 0az** [comment]
It's mostly so we don't have to use hacks like symlinking into `node_modules`, which I *really* want to avoid.

**Comment by jacksonargo** [comment]
Not sure why symlinks are a hack, but does Dockerfile just overwriting the symlink work for you?

**Comment by 0az** [comment]
I guess.

**Pull request merged**